### PR TITLE
Utf8

### PR DESCRIPTION
**Pull request by migueldeicaza**
UTF8 Marshaling APIs  (#3736)

Adds support for the new UTF8 marshaling functionality in .NET, there
is support for the new MarshalAs(UnmanagedType.LPUTF8St) as well as a
handful of convenience methods in Marshal.cs

In Mono, this is mostly a pass-through as we have historically only
done UTF8, so this merely adds the constant processing to the runtime.

This fixes a few bugs that the new test suite exhibited: when we
marshaled UTF8 strings, we assumed Chars() == Bytes() and we would end
up chopping when marshaling out, and marshaling in.

There is also an additional behavioral bug that was fixed in
Marshal.cs, unlike the rest of the runtime that treated Ansi as Utf8,
StringToCoTaskMemAnsi behaved as ascii. This brings the API in line
with the rest of the runtime.

The equivalent CoreCLR changes were:

dotnet/coreclr#4793
dotnet/coreclr#6561

This has also surfaced two limitations in Mono's marshaling code,
which is why two sets of tests are disabled: StringBuilder return
types are not handled specially (char * needs to be turned into a new
StringBuilder when present) and StringBuilders with an Out attribute
should update the StringBuilder in place, not make a new copy of it.

### CODE REVIEW

File: external/ikvm [hunk 1]
```diff
@@ -1 +1 @@
-Subproject commit 367864ef810859ae3ce652864233b35f2dd5fdbe
+Subproject commit 9f44f259bf5059df6c8019948ea75313bd09a9fb
```
Comment (marek-safar) [inline external/ikvm]: Unwanted change

File: mcs/class/corlib/System.Runtime.InteropServices/Marshal.cs [hunk 2]
```diff
@@ -76,6 +77,9 @@ public static bool AreComObjectsAvailableForCleanup ()
 
 		[MethodImplAttribute(MethodImplOptions.InternalCall)]
 		public extern static IntPtr AllocCoTaskMem (int cb);
+		
+		[MethodImplAttribute(MethodImplOptions.InternalCall)]
+		public extern static IntPtr AllocCoTaskMemSize (UIntPtr sizet);
```
Comment (marek-safar) [inline mcs/class/corlib/System.Runtime.InteropServices/Marshal.cs]: Where did you find this new public API?

File: mcs/class/corlib/Test/System.Runtime.InteropServices/MarshalTest.cs [hunk 3]
```diff
@@ -155,6 +155,58 @@ public void PtrToStringWithUni_Ptr_Zero ()
...
+				string retString = Marshal.PtrToStringUTF8(ptrString);
+
+				if (!srcString.Equals(retString))
+				{
+					throw new Exception("Round triped strings do not match...");
+				}
+				if (srcString.Length > 0)
+				{
+					string retString2 = Marshal.PtrToStringUTF8(ptrString, srcString.Length - 1);
+					if (!retString2.Equals(srcString.Substring(0, srcString.Length - 1)))
```
Comment (marek-safar) [inline mcs/class/corlib/Test/System.Runtime.InteropServices/MarshalTest.cs]: Same here

File: mcs/class/corlib/System.Runtime.InteropServices/Marshal.cs (hunk 2 above)
Comment (migueldeicaza) [inline mcs/class/corlib/System.Runtime.InteropServices/Marshal.cs]: My bad, just added this one to tell this one apart from the int version, which is limiting by design.

**Pull request merged**